server/jetstream_api.go

```diff
@@ -1033,6 +1080,193 @@ func (s *Server) jsStreamPurgeRequest(sub *subscription, c *client, subject, rep
+			// TODO(dlc) - no way right now to communicate back.
+			acc.RestoreStream(stream, tfile)
+			tfile.Close()
+			os.Remove(tfile.Name())
+			c.processUnsub(sub.sid)
+			return
+		}
+		// Append chunk to temp file.
+		tfile.Write(msg)
+	})
+
+	resp.DeliverSubject = restoreSubj
+	s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(resp))
+}
+
+// Process a snapshot request.
+func (s *Server) jsStreamSnapshotRequest(sub *subscription, c *client, subject, reply string, msg []byte) {
+	if c.acc == nil {
+		return
+	}
+	acc := c.acc
+
+	var resp = JSApiStreamSnapshotResponse{ApiResponse: ApiResponse{Type: JSApiStreamSnapshotResponseType}}
+	if !acc.JetStreamEnabled() {
+		resp.Error = jsNotEnabledErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+	if isEmptyRequest(msg) {
+		resp.Error = jsBadRequestErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+	stream := streamNameFromSubject(subject)
+	mset, err := acc.LookupStream(stream)
+	if err != nil {
+		resp.Error = jsNotFoundError(err)
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+
+	var req JSApiStreamSnapshotRequest
+	if err := json.Unmarshal(msg, &req); err != nil {
+		resp.Error = jsInvalidJSONErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+	if !IsValidSubject(req.DeliverSubject) {
+		resp.Error = jsBadRequestErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+
+	sr, err := mset.Snapshot(0, !req.NoConsumers)
+	if err != nil {
+		resp.Error = jsError(err)
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+
+	resp.NumBlks = sr.NumBlks
+	resp.BlkSize = sr.BlkSize
+	s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(resp))
+
+	// Now do the real streaming in a separate go routine.
+	go s.streamSnapshot(c, mset, sr, &req)
+}
+
+const defaultSnapshotChunkSize = 64 * 1024
+
+// streamSnapshot will stream out our snapshot to the reply subject.
+func (s *Server) streamSnapshot(c *client, mset *Stream, sr *SnapshotResult, req *JSApiStreamSnapshotRequest) {
+	chunkSize := req.ChunkSize
+	if chunkSize == 0 {
+		chunkSize = defaultSnapshotChunkSize
+	}
+	// Setup for the chunk stream.
+	acc := c.acc
+	reply := req.DeliverSubject
+	r := sr.Reader
+	defer r.Close()
+
+	// Check interest for the snapshot deliver subject.
+	inch := make(chan bool, 1)
+	acc.sl.RegisterNotification(req.DeliverSubject, inch)
+	defer acc.sl.ClearNotification(req.DeliverSubject, inch)
+	hasInterest := <-inch
+	if !hasInterest {
+		// Allow 2 seconds or so for interest to show up.
+		select {
+		case <-inch:
+		case <-time.After(2 * time.Second):
+		}
+	}
+
+	// Create our ack flow handler.
+	// This is very simple for now.
+	acks := make(chan struct{}, 1)
+	acks <- struct{}{}
+
+	ackSubj := fmt.Sprintf("_snapshot_.%s", nuid.Next())
```

Comment:
can we make this `$JS.something` so its clearly associated with the lifecycle and assists with ACLs, acks are `$JS.ACK.>` so maybe `$JS.SNAPSHOT.ACK.>` to give us some space to expand into there for other flow control messages in future - things like client notifying you about windowing etc

Reply:
This would be an allow_response dynamic permission.

Reply:
Not a fixed one IMO.

Reply:
After discussion agree we need better prefixes etc. Fixed.